### PR TITLE
Allow user-supplied secret key backup passphrases

### PR DESCRIPTION
**Pull request by vanitasvitae**
Instead of passing the passphrase back to the user via a
DisplayBackupCodeCallback, we directly return the passphrase
which is now represented by a class.

Also we now allow the user to provide the passphrase
when backing up the key to the server.

### CODE REVIEW

File: smack-core/src/main/java/org/jivesoftware/smack/util/StringUtils.java [hunk 1]
```diff
@@ -270,7 +270,7 @@ public static String encodeHex(byte[] bytes) {
     /**
      * 24 upper case characters from the latin alphabet and numbers without '0' and 'O'.
      */
-    private static final char[] UNAMBIGUOUS_NUMBERS_AND_LETTER = "123456789ABCDEFGHIJKLMNPQRSTUVWXYZ".toCharArray();
+    public static final char[] UNAMBIGUOUS_NUMBERS_AND_LETTER = "123456789ABCDEFGHIJKLMNPQRSTUVWXYZ".toCharArray();
```
Comment (Flowdalic) [inline smack-core/src/main/java/org/jivesoftware/smack/util/StringUtils.java:273]: If this array is public, then users would modify its contents. I think it should stay private. I guess you idea was to use it in the regex below, which we definitely should!
How about:

```
public static final String UNAMBIGUOUS_NUMBERS_AND_LETTER_STRING = "123456789ABCDEFGHIJKLMNPQRSTUVWXYZ";
private static final char[] UNAMBIGUOUS_NUMBERS_AND_LETTER = UNAMBIGUOUS_NUMBERS_AND_LETTER_STRING.toCharArray();
```

File: smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpManager.java [hunk 2]
```diff
@@ -551,7 +577,7 @@ public void run() {
                 if (contentElement instanceof SigncryptElement) {
                     for (SigncryptElementReceivedListener l : signcryptElementReceivedListeners) {
                         l.signcryptElementReceived(contact, message, (SigncryptElement) contentElement,
-                                        decrypted.getMetadata());
```
Comment (Flowdalic) [inline smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpManager.java:554]: Unrelated whitespace change

File: smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpManager.java [hunk 3]
```diff
@@ -566,7 +592,7 @@ public void run() {
                 if (contentElement instanceof CryptElement) {
                     for (CryptElementReceivedListener l : cryptElementReceivedListeners) {
                         l.cryptElementReceived(contact, message, (CryptElement) contentElement,
-                                        decrypted.getMetadata());
+                                decrypted.getMetadata());
```
Comment (Flowdalic) [inline smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpManager.java:595]: Unrelated whitespace change

File: smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpSecretKeyBackupPassphrase.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Represents a secret key backup passphrase whose format is described in XEP-0373 §5.3.
+ *
+ * @see <a href="https://xmpp.org/extensions/xep-0373.html#backup-encryption">
+ *      XEP-0373 §5.4 Encrypting the Secret Key Backup</a>
+ */
+public class OpenPgpSecretKeyBackupPassphrase implements CharSequence {
+
+    private static final Pattern PASSPHRASE_PATTERN = Pattern.compile(
+            "^([123456789ABCDEFGHIJKLMNPQRSTUVWXYZ]{4}-){5}[123456789ABCDEFGHIJKLMNPQRSTUVWXYZ]{4}$");
```
Comment (Flowdalic) [inline smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpSecretKeyBackupPassphrase.java:30]: Should re-use UNAMBIGUOUS_NUMBERS_AND_LETTER_STRING from StringUtils

File: smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpSecretKeyBackupPassphrase.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    private final String passphrase;
+
+    public OpenPgpSecretKeyBackupPassphrase(String passphrase) {
+        if (!PASSPHRASE_PATTERN.matcher(passphrase).matches()) {
```
Comment (Flowdalic) [inline smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpSecretKeyBackupPassphrase.java:35]: I'd love to see unit tests for this constructor. One with a valid passphrase and one with an invalid one would be a great start. :)

File: smack-core/src/main/java/org/jivesoftware/smack/util/StringUtils.java (hunk 1 above)
Comment (vanitasvitae) [reply smack-core/src/main/java/org/jivesoftware/smack/util/StringUtils.java:273]: Ah right, haven't thought of that.
Yeah I guess either the fix you suggested, or just make it private again, as I ended up constructing the regex without the array.

File: smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpManager.java (hunk 3 above)
Comment (vanitasvitae) [reply smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpManager.java:595]: Are you happy if I move this change to another commit?

File: smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpSecretKeyBackupPassphrase.java (hunk 5 above)
Comment (vanitasvitae) [reply smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpSecretKeyBackupPassphrase.java:35]: There is a test for a valid construction already, but I happily add another one as well :)

File: smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpManager.java (hunk 3 above)
Comment (vanitasvitae) [reply smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpManager.java:596]: Moved into separate commit

File: smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpSecretKeyBackupPassphrase.java (hunk 5 above)
Comment (vanitasvitae) [reply smack-openpgp/src/main/java/org/jivesoftware/smackx/ox/OpenPgpSecretKeyBackupPassphrase.java:38]: Added dedicated junit tests.

**Comment by Flowdalic** [comment]
Note that the "Fix indentation" commit message would also be nice if it was prefixed. But since it is just a code style change, I am merging this as is anyways.
